File: app/session/answer-intent.ts
import type { AnswerIntent } from "@/lib/types";

const UNCERTAIN_PATTERNS = [
  /^no idea[.!]?$/i,
  /^i\s+don'?t\s+know\s+yet[.!]?$/i,
  /^i\s+don'?t\s+know[.!]?$/i,
  /^not\s+sure[.!]?$/i,
  /^unsure[.!]?$/i,
];

export const KNOWLEDGE_GAP_ANSWER = "I don't know yet.";

export function answerIntentFor(text: string, intent: AnswerIntent = "attempt"): AnswerIntent {
  const trimmed = text.trim();
  if (intent === "knowledge_gap") return intent;
  return UNCERTAIN_PATTERNS.some((pattern) => pattern.test(trimmed)) ? "knowledge_gap" : "attempt";
}

export function answerTextFor(text: string, intent: AnswerIntent): string {
  return intent === "knowledge_gap" && !text.trim() ? KNOWLEDGE_GAP_ANSWER : text;
}

File: agents/answer_intent.py
from __future__ import annotations

import re

KNOWLEDGE_GAP_ANSWER = "I don't know yet."

_UNCERTAIN_PATTERNS = [
    re.compile(r"^no idea[.!]?$", re.I),
    re.compile(r"^i\s+don'?t\s+know\s+yet[.!]?$", re.I),
    re.compile(r"^i\s+don'?t\s+know[.!]?$", re.I),
    re.compile(r"^not\s+sure[.!]?$", re.I),
    re.compile(r"^unsure[.!]?$", re.I),
]


def normalize_answer_intent(user_answer: str, answer_intent: str = "attempt") -> str:
    if answer_intent == "knowledge_gap":
        return "knowledge_gap"
    answer = user_answer.strip()
    return "knowledge_gap" if any(pattern.match(answer) for pattern in _UNCERTAIN_PATTERNS) else "attempt"
